lib/srv/usermgmt_unix.go

```diff
@@ -0,0 +1,130 @@
+//go:build !windows
+// +build !windows
+
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package srv
+
+import (
+	"os/exec"
+	"os/user"
+	"strings"
+
+	"github.com/gravitational/trace"
+)
+
+/*
+#include <sys/types.h>
+#include <pwd.h>
+*/
+import "C"
+
+// man GROUPADD(8), exit codes section
+const groupExistExit = 9
+
+// man USERADD(8), exit codes section
+const userExistExit = 9
+const userLoggedInExit = 8
+
+type unixMgmt struct{}
+
+var _ UserManagement = &unixMgmt{}
+
+// Lookup implements host user information lookup
+func (*unixMgmt) Lookup(username string) (*user.User, error) {
+	return user.Lookup(username)
+}
+
+func (*unixMgmt) UserGIDs(u *user.User) ([]string, error) {
+	return u.GroupIds()
+}
+
+// LookupGroup host group information lookup
+func (*unixMgmt) LookupGroup(name string) (*user.Group, error) {
+	return user.LookupGroup(name)
+}
+
+func (*unixMgmt) GetAllUsers() ([]string, error) {
+	var result *C.struct_passwd
+	names := []string{}
+	// getpwent(3), posix compatible way to iterate /etc/passwd.
+	// Provided as os/user does not provide any iteration helpers
+	C.setpwent()
+	defer C.endpwent()
+	for {
+		result = C.getpwent()
+		if result == nil {
+			break
+		}
+		name := result.pw_name
+		names = append(names, C.GoString(name))
+	}
+	if len(names) == 0 {
+		return nil, trace.NotFound("failed to find any /etc/passwd entries")
+	}
+	return names, nil
+}
+
+func (*unixMgmt) groupAdd(groupname string) (exitCode int, err error) {
```

Comment:
All these methods that just wrap groupadd/useradd it seems like can live somewhere in utils e.g. `lib/utils/hostusers.go` as utility functions and not be a part of any object.